src/invariant/OrderBookIsNotCrossed.cpp

```diff
@@ -0,0 +1,225 @@
+#ifdef BUILD_TESTS
+// Copyright 2021 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "invariant/OrderBookIsNotCrossed.h"
+#include "invariant/InvariantManager.h"
+#include "ledger/InternalLedgerEntry.h"
+#include "ledger/LedgerTxn.h"
+#include "main/Application.h"
+#include "util/Logging.h"
+#include "util/XDROperators.h"
+#include "xdr/Stellar-ledger-entries.h"
+#include <fmt/format.h>
+#include <functional>
+#include <xdrpp/printer.h>
+
+namespace stellar
+{
+
+namespace
+{
+static AssetPairSet
+extractAssetPairs(LedgerTxnDelta const& ltxd)
+{
+    AssetPairSet assets;
+    for (auto const& entry : ltxd.entry)
+    {
+        if (entry.first.type() == InternalLedgerEntryType::LEDGER_ENTRY &&
+            entry.first.ledgerKey().type() == OFFER && entry.second.current)
+        {
+            auto const& offer =
+                entry.second.current->ledgerEntry().data.offer();
+            auto const assetPair =
+                offer.selling < offer.buying
+                    ? std::make_pair(offer.selling, offer.buying)
+                    : std::make_pair(offer.buying, offer.selling);
+
+            assets.insert(assetPair);
+        }
+    }
+
+    return assets;
+}
+
+static double
+price(OfferEntry const& offer)
+{
+    return double(offer.price.n) / double(offer.price.d);
+}
+
+static std::string
+checkCrossed(Asset const& a, Asset const& b, OrderBook const& orderBook)
+{
+    // if either side of order book for asset pair empty or does not yet exist,
+    // order book cannot be crossed
+    auto const iterAB = orderBook.find({a, b});
+    auto const iterBA = orderBook.find({b, a});
+    if (iterAB == orderBook.end() || iterBA == orderBook.end())
+    {
+        return {};
+    }
+
+    auto const& asks = iterAB->second;
+    auto const& bids = iterBA->second;
+    if (asks.empty() || bids.empty())
+    {
+        return {};
+    }
+
+    // check if crossed
+    auto lowestAsk = price(*asks.cbegin());
+    auto highestBid = 1 / price(*bids.cbegin());
```

Comment:
One thing I am trying to confirm to myself -- or find a counterexample to -- as I consider this is whether we really can always resolve the question of whether the book is crossed by looking at the first offers in the per-asset-pair ask/bid lists.  I see this comment to `exchangeV10()`:

```
// exchangeV10 is a system for crossing offers that provides guarantees
// regarding the direction and magnitude of rounding errors:
// - When considering two crossing offers subject to a variety of limits,
//   exchangeV10 has a consistent approach to determining which offer is larger.
//   The smaller offer is always removed from the book.
// - When two offers cross, the rounding error will favor the offer that remains
//   in the book.
// - The rounding error will not favor either party by more than 1% (except in
//   the case of path payment, where the rounding error can favor the offer in
//   the book by an arbitrary amount). If the rounding error would exceed 1%,
//   no trade occurs and the smaller offer is removed.
```

In particular I think that the "smaller offer is removed" part means that rounding error should not allow the book to cross, but rather, if rounding error would prevent an exchange required to prevent the book from crossing, then one of the offers is simply removed altogether.  And I think that that means in turn that we indeed don't need to search beyond the first element:  I don't think there is a case where we are _expected_ to leave crossed offers in the book because of rounding errors preventing them from being taken.  But I have by no means absorbed all the rules yet.

Reply:
> but suffice it to say that x = 1/49 is such an example

Bingo!  If I write this test:

```
    SECTION("Not crossed where highest bid = lowest ask but only because of "
            "passive offer(s), where double-precision reciprocal of price is "
            "not involutory")
    {
        auto const offer9 = createOffer(cur4, cur5, 1, Price{49, 1}, true);
        auto const offer10 = createOffer(cur5, cur4, 1, Price{1, 49}, true);
        applyCheck(*app, {offer9, offer10}, {}, true);
    }
```

Then I get:

```
2021-03-16T15:08:19.365 <test> [default FATAL] Order book is in a crossed state for { type = ASSET_TYPE_CREDIT_ALPHANUM4,
  alphaNum4 = {
    assetCode = 43555234,
    issuer = {
      type = PUBLIC_KEY_TYPE_ED25519,
      ed25519 = 0000000000000000000000000000000000000000000000000000000000000000
    }
  } }
 - { type = ASSET_TYPE_CREDIT_ALPHANUM4,
  alphaNum4 = {
    assetCode = 43555235,
    issuer = {
      type = PUBLIC_KEY_TYPE_ED25519,
      ed25519 = 0000000000000000000000000000000000000000000000000000000000000000
    }
  } }
 asset pair.
Top of the book is:
        Ask price: 49.0
        Bid price: 49.00000000000001

Where 49.00000000000001 >= 49.0!
```

Reply:
As for fixing it...

`stellar-core` does use `double` when comparing prices, but as far as I can tell, it's in `isBetterOffer` where we're comparing prices with the same ask and bid assets, not with the reversed ones.  When we do a reciprocal, we do it in the rational form:

```
static Price
getInversePrice(Price const& price)
{
    return Price{price.d, price.n};
}
```

And when checking for crossing, we're using the `Price` comparison operators from `util/types.h`:

```
bool
ManageOfferOpFrameBase::doApply(AbstractLedgerTxn& ltxOuter)
{
[ ... ]
                if ((passive && (o.price >= maxWheatPrice)) ||
                    (o.price > maxWheatPrice))
                {
                    return OfferFilterResult::eStop;
                }
[ ... ]
}
```

Those operators use large rational arithmetic:

```
bool
operator>(Price const& a, Price const& b)
{
    uint128_t l(a.n);
    uint128_t r(a.d);
    l *= b.d;
    r *= b.n;
    return l > r;
}
```

So it's not obvious to me why we should be using `double` in the `OrderBookIsNotCrossed` invariant at all.  Indeed, if I switch to using the `Price` comparison operator, then the test above, and a similar one which should throw but doesn't when using `double`, both pass.  So I've committed those tests, and I've also committed changes to remove the use of `double` from the invariant module.

However, there are a couple things which I might want to change but that I'm not sure what the best way would be, so I'll ask here first:

- `types.h` only defines `>`, `>=`, and `==`; it doesn't define `<` or `<=`.  For now I just flipped the `<=` so that I wouldn't have to touch `types.h` in this PR, but I could add it if you prefer.
-  `types.h`'s version of `==` uses structural rather than rational equality:
```
bool
operator==(Price const& a, Price const& b)
{
    return (a.n == b.n) && (a.d == b.d);
}
```
So it's unsuitable for the uses in the invariant module.  In the committed code, I wrote something with the spelled-out name `pricesEqualAsRationals`, which works analogously to `>=` and `>`.  I'm guessing that some things are relying on the current semantics of `types.h`'s `==`, or else it wouldn't be that way, so I won't touch it unless someone can confirm otherwise.  I'm not sure whether I should do something more concise than `pricesEqualAsRationals` or whether the wordiness is actually good for clarity in this case.